### PR TITLE
Feat(eos_designs): Support for custom masks in l3_edge ip pools

### PR DESCRIPTION
**Pull request by philippebureau**
## Change Summary

Add support for custom subnet masks in eos_desings l3_edge ip pools

## Component(s) name

`arista.avd.eos_desings`

## Proposed changes
```yaml
l3_edge:
  p2p_links_ip_pools:
    - name: < p2p_pool_name_1 >
      ipv4_pool_mask: < subnet mask size | default -> 31 >
```

## How to test
See molecule scenario CORE_UNIT_TESTS.yml

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/l3_edge/utils.py [hunk 1]
```diff
@@ -167,11 +167,15 @@ def _resolve_p2p_ips(self, p2p_link: dict) -> dict:
...
+            pool_prefix = get_item(self._p2p_links_ip_pools, "name", p2p_link["ip_pool"], default={}).get("ipv4_pool_mask")
+            if pool_prefix is None:
+                pool_prefix = 31
+
             if not ip_pool:
                 # Not possible to resolve from pool. Returning original
                 return p2p_link
             id = int(p2p_link["id"])
-            subnet = list(islice(ip_network(ip_pool).subnets(new_prefix=31), id - 1, id))[0]
+            subnet = list(islice(ip_network(ip_pool).subnets(new_prefix=int(pool_prefix)), id - 1, id))[0]
```
Comment (gmuloc) [inline ansible_collections/arista/avd/roles/eos_designs/python_modules/l3_edge/utils.py]: Let's simplify this:
```suggestion
            # Resolving subnet from pool
            ip_pool = get_item(self._p2p_links_ip_pools, "name", p2p_link["ip_pool"], default={})
            ip_pool_subnet = ip_pool.get("ipv4_pool")
            if not ip_pool_subnet:
                return p2p_link
            pool_prefix = int(ip_pool.get("ipv4_pool_mask", 31))
            id = int(p2p_link["id"])
            subnet = list(islice(ip_network(ip_pool_subnet).subnets(new_prefix=pool_prefix), id - 1, id))[0]

```

File: ansible_collections/arista/avd/roles/eos_designs/doc/l3-edge.md [hunk 2]
```diff
@@ -11,6 +11,7 @@ Make sure to configure the variables in a group_vars file covering all devices m
 l3_edge:
   p2p_links_ip_pools:
     < p2p_pool_name_1 >: < IPv4_address/Mask >
+      ipv4_pool_mask: < subnet mask size | default -> 31 >
```
Comment (gmuloc) [inline ansible_collections/arista/avd/roles/eos_designs/doc/l3-edge.md]: this looks weird to have both `< IPv4_address/Mask >` and `ipv4_pool_mask` close to each other. Maybe let's use prefix_size ?

Comment (philippebureau) [reply ansible_collections/arista/avd/roles/eos_designs/doc/l3-edge.md]: will change to prefix_size

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/l3_edge/utils.py (hunk 1 above)
Comment (philippebureau) [reply ansible_collections/arista/avd/roles/eos_designs/python_modules/l3_edge/utils.py]: updated with suggestions

File: ansible_collections/arista/avd/roles/eos_designs/doc/l3-edge.md (hunk 2 above)
Comment (philippebureau) [reply ansible_collections/arista/avd/roles/eos_designs/doc/l3-edge.md]: done

File: ansible_collections/arista/avd/roles/eos_designs/python_modules/l3_edge/utils.py [hunk 3]
```diff
@@ -166,12 +166,13 @@ def _resolve_p2p_ips(self, p2p_link: dict) -> dict:
...
         else:
             # Resolving subnet from pool
-            ip_pool = get_item(self._p2p_links_ip_pools, "name", p2p_link["ip_pool"], default={}).get("ipv4_pool")
-            if not ip_pool:
-                # Not possible to resolve from pool. Returning original
+            ip_pool = get_item(self._p2p_links_ip_pools, "name", p2p_link["ip_pool"], default={})
+            ip_pool_subnet = ip_pool.get("ipv4_pool")
+            if not ip_pool_subnet:
                 return p2p_link
+            pool_prefix = int(ip_pool.get("prefix_size", 31))
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/python_modules/l3_edge/utils.py]: please call the var `prefix_size` to make it simpler to follow.

Comment (philippebureau) [reply ansible_collections/arista/avd/roles/eos_designs/python_modules/l3_edge/utils.py]: done

**Review by ClausHolbechArista — APPROVED** [review]

**Review by ClausHolbechArista — APPROVED** [review]
Nice implementation. Thank you for this addition.

**Comment by ClausHolbechArista** [comment]
@philippebureau if you have time, it would be useful to implement the same in core_interfaces. It has 95% identical data model and code.

**Comment by philippebureau** [comment]
> @philippebureau if you have time, it would be useful to implement the same in core_interfaces. It has 95% identical data model and code.

I'll take a look at core_interfaces

File: ansible_collections/arista/avd/roles/eos_designs/doc/l3-edge.md [hunk 4]
```diff
@@ -11,6 +11,7 @@ Make sure to configure the variables in a group_vars file covering all devices m
 l3_edge:
   p2p_links_ip_pools:
     < p2p_pool_name_1 >: < IPv4_address/Mask >
+      prefix_size: < subnet mask size | default -> 31 >
```
Comment (tgodaA) [inline ansible_collections/arista/avd/roles/eos_designs/doc/l3-edge.md]: Invalid data-model.

For example:
```
p2p_links_ip_pools:
  test_pool1: "172.16.0.0/24"
  prefix_size: 30
  test_pool2: 192.168.0.0/24
```

Comment (tgodaA) [reply ansible_collections/arista/avd/roles/eos_designs/doc/l3-edge.md]: The convert_dicts won't convert properly.

**Comment by philippebureau** [comment]
> > @philippebureau if you have time, it would be useful to implement the same in core_interfaces. It has 95% identical data model and code.
> 
> I'll take a look at core_interfaces

PR 2469 submitted for the core_interfaces implementation of the prefix_size
(https://github.com/aristanetworks/ansible-avd/pull/2469)

File: ansible_collections/arista/avd/roles/eos_designs/doc/l3-edge.md (hunk 4 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_designs/doc/l3-edge.md]: we cannot add new keys to the old models, so it is basically not supported in the old input style. But to remember to add this to the schema, I think it's fine to keep it in the doc for now. Schemas will replace the doc before we release 4.0.

Comment (philippebureau) [reply ansible_collections/arista/avd/roles/eos_designs/doc/l3-edge.md]: removed from l2-edge.md

**Review by ClausHolbechArista — APPROVED** [review]

**Review by tgodaA — APPROVED** [review]
LGTM

**Pull request merged**